ogr/ogrsf_frmts/gtfs/ogrgtfsdriver.cpp

```diff
@@ -0,0 +1,724 @@
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include "cpl_port.h"
+#include "gdal_priv.h"
+#include "ogrsf_frmts.h"
+
+#include <map>
+#include <utility>
+
+/***********************************************************************/
+/*                         OGRGTFSDataset                              */
+/***********************************************************************/
+
+class OGRGTFSDataset final : public GDALDataset
+{
+    std::vector<std::unique_ptr<OGRLayer>> m_apoLayers{};
+
+  public:
+    OGRGTFSDataset() = default;
+
+    int GetLayerCount() override
+    {
+        return static_cast<int>(m_apoLayers.size());
+    }
+    OGRLayer *GetLayer(int nIdx) override;
+
+    static int Identify(GDALOpenInfo *poOpenInfo);
+    static GDALDataset *Open(GDALOpenInfo *poOpenInfo);
+};
+
+/***********************************************************************/
+/*                              GetLayer()                             */
+/***********************************************************************/
+
+OGRLayer *OGRGTFSDataset::GetLayer(int nIdx)
+{
+    return nIdx >= 0 && nIdx < static_cast<int>(m_apoLayers.size())
+               ? m_apoLayers[nIdx].get()
+               : nullptr;
+}
+
+/***********************************************************************/
+/*                           OGRGTFSLayer                              */
+/***********************************************************************/
+
+class OGRGTFSLayer final : public OGRLayer
+{
+    std::string m_osDirname{};
+    std::unique_ptr<GDALDataset> m_poUnderlyingDS{};
+    OGRLayer *m_poUnderlyingLayer = nullptr;  // owned by m_poUnderlyingDS
+    OGRFeatureDefn *m_poFeatureDefn = nullptr;
+    int m_nTripIdIdx = -1;
+    int m_nLatIdx = -1;
+    int m_nLonIdx = -1;
+    bool m_bIsTrips = false;
+    bool m_bPrepared = false;
+    std::map<std::string, std::pair<double, double>> m_oMapStopIdToLonLat{};
+    std::map<std::string, std::map<int, std::string>> m_oMapTripIdToStopIds{};
+
+    void PrepareTripsData();
+
+  public:
```

Comment:
I'm not sure about GDAL coding conventions, but I would expect to find the `public` interface first and an explicit `private` secion below it.

Reply:
we don't have that much conventions regarding that. The pattern with private first is actually quite common in GDAL sources